core/lib/contract_verifier/src/lib.rs

```diff
@@ -353,6 +364,99 @@ impl ContractVerifier {
         Ok((info, compiled_identifier))
     }
 
+    /// Returns compilation artifacts for the given request.
+    /// Tries to use the compiler versions from the metadata if they are different from the ones in the request.
+    /// If the metadata is not present, or the versions are the same, uses the ones from the request.
+    async fn get_compilation_artifacts(
+        &self,
+        request: &mut VerificationRequest,
+        detected_metadata: Option<&DetectedMetadata>,
+        bytecode_marker: &BytecodeMarker,
+        deployed_code: &[u8],
+    ) -> Result<CompilationArtifacts, ContractVerifierError> {
+        match detected_metadata {
+            Some(DetectedMetadata::Cbor {
+                metadata: cbor_metadata,
+                ..
+            }) if !request.req.compiler_versions_match(cbor_metadata) => {
+                let updated_request = request
+                    .req
+                    .clone()
+                    .with_updated_compiler_versions(cbor_metadata);
+
+                match self
+                    .compile(updated_request.clone(), *bytecode_marker)
+                    .await
+                {
+                    Ok(artifacts) if artifacts.match_bytecode(bytecode_marker, deployed_code) => {
```

Comment:
Under the hood you calculate identifier, but you have it already: it can be passed instead of `detected_metadata`
(and I guess it'll make `match_bytecode` method redundant).